client/tinychain/ml/linalg.py

```diff
@@ -273,3 +277,235 @@ def svd(matrix: Tensor) -> Tuple:
+                @closure(U, V, e, txn.m, txn.n)
+                @post_op
+                def step(j: UInt, l: UInt, i: UInt) -> Map:
+                    cxt.col_j = V[l:, j]
+                    return Map(After(
+                        cxt.col_j.write(cxt.col_j + V[l:, i] * dot(U[i, l:], cxt.col_j)),
+                        Map(j=j + 1, n=txn.n, l=l, i=i, U=U, V=V)
+                    ))
+
+                @closure(txn.n)
+                @post_op
+                def cond(j: UInt) -> Bool:
+                    return j < txn.n
+
+                loop = After(V[l:,i].write((U[i,l:] / U[i,l]) / g), While(cond, step, Map(j=l, n=txn.n, l=l, i=i, U=U, V=V)))
+
+                return loop
+
+            cxt.update_cols = update_cols
+            cxt.res = cxt.update_cols(n=txn.n, l=l, i=i, U=U,  V=V)
+            return If(Bool(g != 0.0), cxt.res)
+
+        cxt.update_g = Float(e[i].copy())
+        cxt.update_cols_wo_last = update_cols_wo_last
+        cxt.update_cols_funct = After(cxt.update_cols_wo_last(l=l, g=g), [V[i,l:].write(V[i,l:]*0.0), V[l:,i].write(V[l:,i]*0.0)])
+        
+        cxt.check_last_col = If(
+            Bool(i < txn.n-1),
+            cxt.update_cols_funct)
+        
+        g_final = After([cxt.check_last_col, V[i,i].write(1.0)], cxt.update_g)
+
+        return Map(i=i - 1, l=i, g=g_final, U=U, V=V, e=e)
+
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i >= 0
+
+    return Map(While(cond, step, Map(i=txn.n-1, l=0, g=0.0, U=U, V=V, e=e)))
+
+
+@post_op
+def lht(txn, U: Tensor, W: Tensor) -> Map:
+    """Accumulation of left hand transformations (lht)"""
+
+    txn.shape = U.shape
+    txn.m, txn.n = [UInt(dim) for dim in txn.shape.unpack(2)]
+    txn.min_m_n = UInt(If(txn.m < txn.n, txn.m, txn.n))
+    
+    @closure(U, W, txn.m, txn.n)
+    @post_op
+    def step(cxt, i: UInt, l: UInt, g: F32) -> Map:
+        
+        @closure(U, W, txn.m, txn.n)
+        @post_op
+        def update_cols(cxt, i: UInt, l: UInt, g: F32) -> F32:
+
+            @closure(U, W, txn.m, txn.n)
+            @post_op
+            def step(j: UInt, l: UInt, i: UInt) -> Map:
+                cxt.col_j = U[i:,j]
+                return Map(After(
+                    cxt.col_j.write(cxt.col_j + U[i:, i] * ((dot(U[l:, i], U[l:, j])/U[i,i])*g)),
+                    Map(j=j + 1, n=txn.n, l=l, i=i, U=U, W=W)
+                ))
+
+            @closure(txn.n)
+            @post_op
+            def cond(j: UInt) -> Bool:
+                return j < txn.n
+
+            cxt.update_g = 1.0 / g
+            loop = After(cxt.update_g, While(cond, step, Map(j=l, n=txn.n, l=l, i=i, U=U, W=W)))
+            update_U = After([loop, U[i:,i].write(U[i:,i]*cxt.update_g)], cxt.update_g)
+
+            return If(Bool(g != 0.0), update_U, g)
+
+        cxt.update_cols = update_cols
+        cxt.new_l = UInt(i + 1)
+        cxt.l = UInt(After(U[i, cxt.new_l:].write(U[i, cxt.new_l:]*0), cxt.new_l))
+        cxt.new_g = Float(W[i].copy())
```

Comment:
`copy` is only defined on `Tensor` so either this should not be a `Float` or you don't need the `copy`